trajopt_ifopt/src/cartesian_line_constraint.cpp

```diff
@@ -0,0 +1,239 @@
+ * @par
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <trajopt_ifopt/constraints/cartesian_line_constraint.h>
+#include <trajopt_ifopt/utils/numeric_differentiation.h>
+
+TRAJOPT_IGNORE_WARNINGS_PUSH
+#include <tesseract_kinematics/core/utils.h>
+#include <console_bridge/console.h>
+TRAJOPT_IGNORE_WARNINGS_POP
+
+namespace trajopt
+{
+CartLineConstraint::CartLineConstraint(const Eigen::Isometry3d& origin_pose,
+                                      const Eigen::Isometry3d& target_pose,
+                                      CartPosKinematicInfo::ConstPtr kinematic_info,
+                                      JointPosition::ConstPtr position_var,
+                                      const std::string& name)
+  : ifopt::ConstraintSet(6, name)
+  , position_var_(std::move(position_var))
+  , line_start_(origin_pose)
+  , line_end_(target_pose)
+  , kinematic_info_(std::move(kinematic_info))
+{
+  // Set the n_dof and n_vars for convenience
+  n_dof_ = kinematic_info_->manip->numJoints();
+  assert(n_dof_ > 0);
+
+  bounds_ = std::vector<ifopt::Bounds>(6, ifopt::BoundZero);
+
+  // calculate the equation of the constraint line
+  line_  = line_end_.translation() - line_start_.translation();//<< line_end_.translation().x() - line_start_.translation().x(), line_end_.translation().y() - line_start_.translation().y(), line_end_.translation().z() - line_start_.translation().z();
+
+}
+
+Eigen::VectorXd CartLineConstraint::CalcValues(const Eigen::Ref<const Eigen::VectorXd>& joint_vals)
+{
+
+  Eigen::Isometry3d new_pose;
+  kinematic_info_->manip->calcFwdKin(new_pose, joint_vals, kinematic_info_->kin_link->link_name);
+
+  new_pose = kinematic_info_->world_to_base * new_pose * kinematic_info_->kin_link->transform * kinematic_info_->tcp;
+
+  // For Jacobian Calc, we need the inverse of the nearest point, D, to new Pose, C, on the constraint line AB
+  // This will not honor tcp orientation. Not used bc this returns a scalar error, need vector/direction components
+//  Eigen::Vector3d d1,d2, pose_err;
+
+  //AC, new pose to first point on line
+//  d1 = new_pose.translation() - line_start_.translation();
+  //BC, new pose to second point on line
+//  d2 = new_pose.translation() - line_end_.translation();
+  //Distance = |AC * BC| / |AB|
+
+//  pose_err = d1.cross(d2).squaredNorm() /  line_.squaredNorm();
+
+  //Find D, the closest point on the line to C, the new point
+
+  //AC, new pose to first point on line
+  Eigen::Vector3d d1 = new_pose.translation() - line_start_.translation();
+
+  // ((AB * AC ) / (|AB|^2)) * (AB)
+  Eigen::Isometry3d temp_ = Eigen::Isometry3d::Identity();
+  double line_point_dist_ = (line_.dot(d1) / line_.dot(line_));
```

Comment:
Maybe clean up the comments here. I can follow it since I have had it explained to me, but I'm not sure it would be readily apparent what/why you were doing this otherwise.

e.g. Project current tcp position onto line and calculate its normalized distance along the line for orientation error.